### DISCUSSION TITLE
`set_global_default` use

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/tracing#1641] [Q&A] [answered]
If I move the execution of `set_global_default` from `main()` into another function, it ceases registering trace events.  How should it be correctly done?

**Comment by foresterre** [comment, marked as answer]
Without the code this is difficult to answer, however it may be that a guard returned from a writer does not live long enough. By moving the guard into another function `f`, it will at most live as long as `f` and be dropped at the end of the scope of `f`. If that's the case, you may want to look at this answer: https://github.com/tokio-rs/tracing/discussions/1494#discussioncomment-1139999

Driving into the code behind `set_global_default` [1], I would not expect this to occur because of moving _just_ `set_global_default`. The `set_global_default` set the default `Dispatch` as a static field for the whole process.

[1] https://github.com/tokio-rs/tracing/blob/a1868eacd606e0922bfff492995258572c123a76/tracing-core/src/dispatch.rs#L302

**Reply by kellytk** [reply]
Thank you for the explanation and pointer.  Returning `WorkerGuard` from the function and persisting it in `main()` with

    let _ = init_tracing();

was indeed the solution.
